src/fsharp/MethodCalls.fs

```diff
@@ -1183,6 +913,408 @@ let BuildMethodCall tcVal g amap isMutable m isProp minfo valUseFlags minst objA
+                        else 
+                            // CSharpMethod(?x=b) when 'b' has optional type and 'x' has non-nullable type --> CSharpMethod(x=Option.defaultValue DEFAULT v)
+                            let _wrapper, defaultExpr = GetDefaultExpressionForCallerSideOptionalArg tcFieldInit g calledArg calledArgTy dfltVal eCallerMemberName m
+                            let ty = destOptionTy g callerArgTy
+                            mkOptionDefaultValue g m ty defaultExpr callerArgExpr
+                    else
+                        // This should be unreachable but the error will be reported elsewhere
+                        callerArgExpr
+                else
+                    if isNullableTy g calledArgTy  then 
+                        // CSharpMethod(x=b) when 'x' has nullable type
+                        if isNullableTy g callerArgTy then 
+                            // CSharpMethod(x=b) when both 'x' and 'b' have nullable type --> CSharpMethod(x=b)
+                            callerArgExpr
+                        else
+                            // CSharpMethod(x=b) when 'x' has nullable type and 'b' does not --> CSharpMethod(x=Nullable(b))
+                            let calledNonOptTy = destNullableTy g calledArgTy 
+                            let minfo = GetIntrinsicConstructorInfosOfType infoReader m calledArgTy |> List.head
+                            let callerArgExprCoerced = mkCoerceIfNeeded g calledNonOptTy callerArgTy callerArgExpr
+                            MakeMethInfoCall amap m minfo [] [callerArgExprCoerced]
+                    elif isOptionTy g calledArgTy then 
+                        // CSharpMethod(x=b) when 'b' has nullable type and 'x' has optional type --> CSharpMethod(Some b.Value)
+                        let calledNonOptTy = destOptionTy g calledArgTy 
+                        let callerArgExprCoerced = mkCoerceIfNeeded g calledNonOptTy callerArgTy callerArgExpr
+                        mkSome g (destNullableTy g callerArgTy) callerArgExprCoerced m
+                    else 
+                        // CSharpMethod(x=b) --> CSharpMethod(?x=b)
+                        callerArgExpr
+
+            | CalleeSide -> 
+                if isOptCallerArg then 
+                    // CSharpMethod(?x=b) --> CSharpMethod(?x=b)
+                    callerArgExpr 
+                else                            
+                    // CSharpMethod(x=b) when CSharpMethod(A) --> CSharpMethod(?x=Some(b :> A))
+                    let calledArgTy = assignedArg.CalledArg.CalledArgumentType
+                    if isOptionTy g calledArgTy then 
+                        let calledNonOptTy = destOptionTy g calledArgTy 
+                        let callerArgExprCoerced = mkCoerceIfNeeded g calledNonOptTy callerArgTy callerArgExpr
+                        mkSome g calledNonOptTy callerArgExprCoerced m
+                    else 
+                        callerArgExpr // should be unreachable 
```

Comment:
Is the "should be unreachable" in the above line an invariant or an aspiration?
probably should be a failwith or remove the comment.

Reply:
thanks, I added an assert to the branch